Wrappers/Python/test/test_CIL_CVXPy.py

```diff
@@ -0,0 +1,148 @@
+
+    
+    # Create gradient operator as sparse matrix that will be used in CVXpy
+    def sparse_gradient_matrix(self, shape, direction='forward', order=1, boundaries='Neumann', **kwargs):
+        
+        len_shape = len(shape)    
+        allMat = dict.fromkeys(range(len_shape))
+        discretization = kwargs.get('discretization',[1.0]*len_shape)
+
+        if order == 1:
+
+            for i in range(0,len_shape):
+
+                if direction == 'forward':
+
+                    mat = 1/discretization[i] * sp.spdiags(np.vstack([-np.ones((1,shape[i])),np.ones((1,shape[i]))]), [0,1], shape[i], shape[i], format = 'lil')
+
+                    if boundaries == 'Neumann':
+                        mat[-1,:] = 0
+                    elif boundaries == 'Periodic':
+                        mat[-1,0] = 1
+
+                elif direction == 'backward':
+
+                    mat = 1/discretization[i] * sp.spdiags(np.vstack([-np.ones((1,shape[i])),np.ones((1,shape[i]))]), [-1,0], shape[i], shape[i], format = 'lil')
+
+                    if boundaries == 'Neumann':
+                        mat[:,-1] = 0
+                    elif boundaries == 'Periodic':
+                        mat[0,-1] = -1
+
+                tmpGrad = mat if i == 0 else sp.eye(shape[0])
+
+                for j in range(1, len_shape):
+
+                    tmpGrad = sp.kron(mat, tmpGrad ) if j == i else sp.kron(sp.eye(shape[j]), tmpGrad )
+
+                allMat[i] = tmpGrad
+
+        else:
+            raise NotImplementedError    
+
+        return allMat        
+
+    def tv_cvxpy_regulariser(self, u, isotropic=True, direction = "forward", boundaries = "Neumann"):
+
+        G = self.sparse_gradient_matrix(u.shape, direction = direction, order = 1, boundaries = boundaries)   
+
+        DX, DY = G[1], G[0]
+    
+        if isotropic:
+            return cp.sum(cp.norm(cp.vstack([DX @ cp.vec(u), DY @ cp.vec(u)]), 2, axis = 0))
+        else:
+            return cp.sum(cp.norm(cp.vstack([DX @ cp.vec(u), DY @ cp.vec(u)]), 1, axis = 0)) 
+
+    def test_cil_vs_cvxpy_totalvariation_isotropic(self):
+
+        # solution
+        u_cvx = cp.Variable(self.data.shape)
+
+        # regularisation parameter
+        alpha = 0.1
+
+        # fidelity term
+        fidelity = 0.5 * cp.sum_squares(u_cvx - self.data.array)   
+        regulariser = alpha * self.tv_cvxpy_regulariser(u_cvx)
+
+        # objective
+        obj =  cp.Minimize( regulariser +  fidelity)
+        prob = cp.Problem(obj, constraints = [])
+
+        # Choose solver ( SCS, MOSEK(license needed) )
+        tv_cvxpy = prob.solve(verbose = True, solver = cp.SCS)        
+
+        # use TotalVariation from CIL (with Fast Gradient Projection algorithm)
+        TV = alpha * TotalVariation(max_iteration=200)
+        tv_cil = TV.proximal(self.data, tau=1.0)     
+
+        # compare solution
+        np.testing.assert_allclose(tv_cil.array, u_cvx.value,atol=1e-3)   
```

Comment:
OK, so this tests that the proximal of the TotalVaration function matches the solution given by CVX, which is great.  I think we should add a simple check that the function value itself also matches what is computed using the sparse difference matrices used by cvxpy.